homeassistant/components/obihai/__init__.py

```diff
@@ -13,6 +15,32 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     return True
 
 
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Migrate old entry."""
+
+    version = entry.version
+
+    LOGGER.debug("Migrating from version %s", version)
+    if version != 2:
+        requester = ObihaiConnection(
+            entry.data[CONF_HOST],
+            username=entry.data[CONF_USERNAME],
+            password=entry.data[CONF_PASSWORD],
+        )
+        await hass.async_add_executor_job(requester.update)
+
+        new_unique_id = await hass.async_add_executor_job(
+            requester.pyobihai.get_device_mac
+        )
+        if entry.unique_id != new_unique_id:
+            entry.version = 2
```

Comment:
Why are we bumping the version conditionally?

Reply:
Good point, I was intending to postpone the version update until the very last second, just in case anything goes wrong with the migration, then at least it'll try again next update.  I'll move this somewhere more intelligent.

Just a point of clarity, because I'm failing to find this anywhere: If I wait until the end of the function, will I need to run async_update_entry() again?  Seems like some components do it, and others don't.

Reply:
> If I wait until the end of the function, will I need to run async_update_entry() again? Seems like some components do it, and others don't.

No, that's not needed. If `async_migrate_entry` returns `True`, config entry data will be saved, including the version.
This PR removes such calls: https://github.com/home-assistant/core/pull/90575